### ISSUE TITLE
Enterprise xml best programming agile practices patterns

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#4321]
Enterprise xml best programming agile practices patterns<br>
<br>
http://bit.ly/2R4XQiX<br>
<br>
About 0 Minutes<br>
<br>
<div><div>
<div>
<div>
<article>
<p>This post is part of the blogpost series explaining coroutines, how they implemented in various programming languages and how they can make your life better:</p>
<ol>
<li><a href="https://dkandalov.github.io/coroutines-as-threads">coroutines as threads</a></li>
<li><a href="https://dkandalov.github.io/yielding-generators">yielding generators</a></li>
<li>👉 <a href="https://dkandalov.github.io/async-await">async await</a> 👈</li>
<li><a href="https://dkandalov.github.io/call-with-current-continuation">call with current continuation</a></li>
</ol>

<p>This post is about <code>async/await</code> coroutines implementation which is a bit more complicated than <code>yield</code> but fundamentally, just like other coroutine implementations, uses the idea of suspending functions and continuing execution some time later from the same point. So the current thread doesn’t really “await” for anything but rather suspends execution of async function.</p>
<h4>Why use async/await?</h4>
<p>The main use-case for <code>async/await</code> is to make it easier to write asynchronous code or, in general, any code which uses a lot of callbacks or <a href="https://en.wikipedia.org/wiki/Futures_and_promises">futures/promises</a>. In particular, <code>async/await</code> can help to avoid <a href="http://callbackhell.com">callback hell</a> (aka <a href="https://en.wikipedia.org/wiki/Pyramid_of_doom_(programming)">pyramid of doom</a>) by writing programs in imperative style. The goal is to restructure program so that it’s easier for humans to write and maintain. Even though asynchronous programs is the most common use-case, <code>async/await</code> doesn’t require code to be actually asynchronous or use non-blocking IO. Fundamentally, it’s just about structural transformation of program.</p>
<h4>Control flow of async/await</h4>
<p>One of the simplest <code>async/await</code> examples is an async function without any return values which <code>await</code>s for two promises. Here <code>promise</code> means an object which holds the result of a computation which might be computed and will become available some time in the future (there are <a href="https://en.wikipedia.org/wiki/Futures_and_promises">other terms</a> for it, e.g. future, deferred or task). This is depicted in the following diagram (you can find description of the <a href="https://dkandalov.github.io/coroutines-as-threads#notation">notation in the previous blogpost</a>):</p>
<p><img src="https://dkandalov.github.io/assets/images/coroutines/async-await/0-async-await-no-return.png"></p>
<p>In the diagram a thread starts executing <code>main</code> and calls async function. Few instructions later async function reaches <code>await</code> which, assuming that <code>promise1</code> is not yet complete, suspends execution of async function and yields execution back to <code>main</code>. Note that there is no restriction on when and how the promise was created. The most common use-case would be to call from async function some long-running task which returns a promise but we could also create the promise some time before and pass into async function as an argument. After <code>await</code> yielded execution back to <code>main</code>, the thread executes the rest of the <code>main</code> and returns from it as usual.</p>
<p>Some time later another thread observes (e.g. via non-blocking IO) that the long-running task has finished and completes <code>promise1</code>. Completing <code>promise1</code> makes thread to “jump” back into the middle of the async function and continue from the <strong>suspension point</strong>. Few instructions later the thread reaches the second <code>await</code> which assuming that <code>promise2</code> is not yet complete, yields execution back to whatever the thread was doing before completing <code>promise1</code>.</p>
<p>Some time later yet another thread observes that the second long-running task has finished so it completes <code>promise2</code> and resumes execution of the async function from the last <strong>suspension point</strong>. This time the thread finishes execution of the whole async function. Even though there were three threads mentioned in this description, coroutines don’t mandate any particular threading model, and in this example there could be any amount of threads between one and three.</p>
<p>One of the most confusing things is that <code>await</code> doesn’t really make current thread wait for anything. If the promise is not yet completed, <code>await</code> chains execution of the rest of async function to run after the promise is completed (e.g. by saving current stack and instruction pointer at suspension point) and yields execution back to the calling function. If the promise is already completed, <code>await</code> extracts result from the promise and continues executing async function as usual. The confusion happens because we normally think about code from the point of view of a computer (CPU or a thread) sequentially executing instructions. When we see <code>sleep(2)</code>, we think of it as current thread sleeping for two seconds (or milliseconds). Unlike most of the normal code, <code>await</code> keyword means “waiting” from the point of view of the current task.</p>
<p>The following JavaScript code shows <code>async/await</code> execution similar to the diagram above:</p>
<div><div><pre><code><span>let</span> promise1 = <span>new</span> <span>Promise</span>(<span><span>resolve</span> =&gt;</span> {
    setTimeout(<span><span>()</span> =&gt;</span> resolve(), <span>2000</span>);
});
<span>let</span> promise2 = <span>new</span> <span>Promise</span>(<span><span>resolve</span> =&gt;</span> {
    setTimeout(<span><span>()</span> =&gt;</span> resolve(), <span>5000</span>);
});

<span>async</span> <span><span>function</span> <span>c</span>(<span></span>) </span>{
    <span>console</span>.log(<span>2</span>);
    <span>await</span> promise1;
    <span>console</span>.log(<span>4</span>);
    <span>await</span> promise2;
    <span>console</span>.log(<span>5</span>);
}

<span>console</span>.log(<span>1</span>);
c();
<span>console</span>.log(<span>3</span>); 
</code></pre></div></div>
<p>Note that <code>function c()</code> is marked with <code>async</code> keyword which tells JavaScript interpreter that this is a coroutine (similar to <a href="https://dkandalov.github.io/yielding-generators#basic-generator">generators</a> which in JavaScript are marked with <code>*</code>). Both <code>promise1</code> and <code>promise2</code> signify some long-running tasks which in this example are resolved 2 and 5 seconds after being created. It also might be a bit counter-intuitive that <code>main</code> finishes execution before async function <code>c</code>. As mentioned above this is because <code>await</code> doesn’t wait for promise to complete and return to <code>main</code>. As you probably expected, the program prints:</p>
<p>Waits for about 2 seconds and prints:</p>
<p>Waits for about (5 - 2) = 3 seconds and prints:</p>
<h4>Returning async results</h4>
<p>The goal of the previous example was to show execution flow of async functions. However, it was simplified by having async function which ignored results of the promises and didn’t return anything itself. This example adds more details about how async functions return values via <a href="https://en.wikipedia.org/wiki/Futures_and_promises">promises</a>. The following diagram shows async function which <code>await</code>s for two promises. After the first <code>await</code> async function returns <code>overall promise</code> to the <code>main</code> which can potentially chain more code to be executed once <code>overall promise</code> is completed. After both <code>promise1</code> and <code>promise2</code> are completed, the thread that completed <code>promise2</code> will execute the rest of async function and complete <code>overall promise</code>.</p>
<p><img src="https://dkandalov.github.io/assets/images/coroutines/async-await/1-async-await.png"></p>
<p>There are some details though which are hard to show on a diagram and are better illustrated with source code:</p>
<div><div><pre><code><span>let</span> promise1 = <span>new</span> <span>Promise</span>(<span><span>resolve</span> =&gt;</span> {
    setTimeout(<span><span>()</span> =&gt;</span> resolve(<span>"🐶"</span>), <span>2000</span>);
});
<span>let</span> promise2 = <span>new</span> <span>Promise</span>(<span><span>resolve</span> =&gt;</span> {
    setTimeout(<span><span>()</span> =&gt;</span> resolve(<span>"🐷"</span>), <span>5000</span>);
});

<span>async</span> <span><span>function</span> <span>c</span>(<span></span>) </span>{
    <span>let</span> dog = <span>await</span> promise1;
    <span>let</span> pig = <span>await</span> promise2;
    <span>return</span> dog + pig;
}

<span>let</span> overallPromise = c();
overallPromise.then(<span><span>result</span> =&gt;</span> <span>console</span>.log(result + <span>"!"</span>));
</code></pre></div></div>
<p>In this example, <code>promise1</code> returns <code>🐶</code> after 2 second delay and <code>promise2</code> returns <code>🐷</code> after 5 seconds. The async function sequentially <code>await</code>s for each promise to complete and returns concatenated output (after about 5 second delay). Note that even though the async function returns a string, actual object received by <code>main</code> is a promise (this happens because <code>function c()</code> has <code>async</code> modifier which tells JavaScript it should do some code transformations). In the <code>main</code> function we register callback on <code>overallPromise</code> to append “!” to the <code>result</code> and print it to the console. As expected, the program runs for about 5 seconds and prints:</p>
<h4>Composing async functions</h4>
<p>It might be obvious, but it’s still worth mentioning that async functions <code>await</code> on promises and return promises themselves. Because of this async functions can be easily composed. For example, in the diagram below async function sequentially invokes <code>asyncFunction1()</code>, <code>asyncFunction2()</code> and <code>await</code>s on their promises.</p>
<p><img src="https://dkandalov.github.io/assets/images/coroutines/async-await/2-async-await-composed.png"></p>
<p>Equivalent JavaScript code might look something like this:</p>
<div><div><pre><code><span>let</span> promise1 = <span>new</span> <span>Promise</span>(<span><span>resolve</span> =&gt;</span> {
    setTimeout(<span><span>()</span> =&gt;</span> resolve(<span>"🐶"</span>), <span>2000</span>);
});
<span>let</span> promise2 = <span>new</span> <span>Promise</span>(<span><span>resolve</span> =&gt;</span> {
    setTimeout(<span><span>()</span> =&gt;</span> resolve(<span>"🐷"</span>), <span>5000</span>);
});

<span>async</span> <span><span>function</span> <span>dog</span>(<span></span>) </span>{
    <span>return</span> <span>await</span> promise1 + <span>"_"</span>;
}
<span>async</span> <span><span>function</span> <span>pig</span>(<span></span>) </span>{
    <span>return</span> <span>await</span> promise2 + <span>"_"</span>;
}
<span>async</span> <span><span>function</span> <span>c</span>(<span></span>) </span>{
    <span>return</span> <span>await</span> dog() + <span>await</span> pig();
}

<span>let</span> overallPromise = c();
overallPromise.then(<span><span>result</span> =&gt;</span> <span>console</span>.log(result + <span>"!"</span>));
</code></pre></div></div>
<p>The main point here is that <code>c()</code> awaits on the result of another async function which could also depend on other async functions. After about 5 second delay the program prints:</p>
<h4>Async try/catch</h4>
<p>Considering previous examples it might look like <code>async/await</code> transforms code between <code>await</code> keywords into promises and chains the promises together. This is not entirely true though. In the example below you can see that <a href="https://dkandalov.github.io/yielding-generators#try-catch">similar to generators</a>, <code>async</code> functions work with <code>try/catch/finally</code> blocks.</p>
<div><div><pre><code><span>let</span> promise1 = <span>new</span> <span>Promise</span>(<span><span>resolve</span> =&gt;</span> {
    setTimeout(<span><span>()</span> =&gt;</span> resolve(<span>"🐶"</span>), <span>200</span>);
});
<span>let</span> promise2 = <span>new</span> <span>Promise</span>(<span>(<span>resolve, reject</span>) =&gt;</span> {
    setTimeout(<span><span>()</span> =&gt;</span> reject(<span>"💥"</span>), <span>500</span>);
});

<span>async</span> <span><span>function</span> <span>c</span>(<span></span>) </span>{
    <span>try</span> {
        <span>let</span> value1 = <span>await</span> promise1;
        <span>let</span> value2 = <span>await</span> promise2;
        <span>console</span>.log(<span>"🙈"</span>);
        <span>return</span> value1 + value2;
    } <span>catch</span> (e) {
        <span>console</span>.log(<span>"catch "</span> + e);
    } <span>finally</span> {
        <span>console</span>.log(<span>"finally"</span>);
    }
}

<span>let</span> overallPromise = c();
overallPromise.then(<span><span>result</span> =&gt;</span> <span>console</span>.log(result));
</code></pre></div></div>
<p>In this example <code>promise1</code> returns <code>🐶</code>, and <code>promise2</code> throws <code>💥</code>. After a delay <code>promise1</code> will be evaluated in async function <code>c()</code> to <code>🐶</code> and assigned to <code>value1</code>. However, <code>await promise2</code> will complete by throwing <code>💥</code> and will continue by executing <code>catch</code> and <code>finally</code> blocks (the code which logs <code>🙈</code> to console and returns concatenated values will never run). Note that even though async function yields execution twice, the <code>finally</code> block is run only once when the whole async function is finished. Because <code>catch/finally</code> doesn’t return any values <code>overallPromise</code> result will be <code>undefined</code>. Overall, the program prints:</p>
<div><div><pre><code><span>catch</span> 💥
<span>finally</span>
<span>undefined</span>
</code></pre></div></div>
<h4>Async/await as state machine</h4>
<p>It might seem that <code>async/await</code> does some complicated compiler/interpreter magic. Even though it’s not trivial there nothing magical about it. Using <a href="https://babeljs.io">BabelJS</a> we can <a href="https://en.wikipedia.org/wiki/Source-to-source_compiler">transpile</a> the code from previous example to be compatible with older JavaScript versions without <code>async/await</code> support.</p>
<p>You don’t need to understand all the code below (and parts of it are defined in BabelJS library anyway), the point here is that, <a href="https://dkandalov.github.io/yielding-generators#state-machines">similar to generators</a>, async function is transformed into a state machine. The original code is split into snippets along <code>await</code>, <code>catch</code> and <code>finally</code> keywords. State matching does a <code>switch</code> on the current state, executes the appropriate snippet and moves to the next state. You might also notice <code>_asyncToGenerator()</code> function which hints that under the hood async functions are implemented as generators, i.e. fundamentally async functions can be expressed using generators.</p>
<div><div><pre><code><span>"use strict"</span>; 
<span>var</span> c = <span><span>function</span> (<span></span>) </span>{
    <span>var</span> _ref = _asyncToGenerator( regeneratorRuntime.mark(<span><span>function</span> <span>_callee</span>(<span></span>) </span>{
        <span>var</span> value1, value2;
        <span>return</span> regeneratorRuntime.wrap(<span><span>function</span> <span>_callee$</span>(<span>_context</span>) </span>{
            <span>while</span> (<span>1</span>) {
                <span>switch</span> (_context.prev = _context.next) {
                    <span>case</span> <span>0</span>:
                        _context.prev = <span>0</span>;
                        _context.next = <span>3</span>;
                        <span>return</span> promise1;

                    <span>case</span> <span>3</span>:
                        value1 = _context.sent;
                        _context.next = <span>6</span>;
                        <span>return</span> promise2;

                    <span>case</span> <span>6</span>:
                        value2 = _context.sent;

                        <span>console</span>.log(<span>"🙈"</span>);
                        <span>return</span> _context.abrupt(<span>"return"</span>, value1 + value2);

                    <span>case</span> <span>11</span>:
                        _context.prev = <span>11</span>;
                        _context.t0 = _context[<span>"catch"</span>](<span>0</span>);

                        <span>console</span>.log(<span>"catch "</span> + _context.t0);

                    <span>case</span> <span>14</span>:
                        _context.prev = <span>14</span>;

                        <span>console</span>.log(<span>"finally"</span>);
                        <span>return</span> _context.finish(<span>14</span>);

                    <span>case</span> <span>17</span>:
                    <span>case</span> <span>"end"</span>:
                        <span>return</span> _context.stop();
                }
            }
        }, _callee, <span>this</span>, [[<span>0</span>, <span>11</span>, <span>14</span>, <span>17</span>]]);
    }));

    <span>return</span> <span><span>function</span> <span>c</span>(<span></span>) </span>{
        <span>return</span> _ref.apply(<span>this</span>, <span>arguments</span>);
    };
}();

<span><span>function</span> <span>_asyncToGenerator</span>(<span>fn</span>) </span>{ ... }

<span>var</span> promise1 = <span>new</span> <span>Promise</span>(<span><span>function</span> (<span>resolve</span>) </span>{
    setTimeout(<span><span>function</span> (<span></span>) </span>{
        <span>return</span> resolve(<span>"🐶"</span>);
    }, <span>100</span>);
});
<span>var</span> promise2 = <span>new</span> <span>Promise</span>(<span><span>function</span> (<span>resolve, reject</span>) </span>{
    setTimeout(<span><span>function</span> (<span></span>) </span>{
        <span>return</span> reject(<span>"💥"</span>);
    }, <span>200</span>);
});

<span>var</span> overallPromise = c();
overallPromise.then(<span><span>function</span> (<span>result</span>) </span>{
    <span>return</span> <span>console</span>.log(result);
});
</code></pre></div></div>
<h4>Summary</h4>
<p>Coroutines based on <code>async/await</code> are probably the most complicated to understand compared to other coroutine implementations. Fundamentally, <code>async/await</code> uses the same idea as other coroutines of saving current stack and executions pointer and later using this information to continue execution from suspension point. Unlike other coroutines, <code>async/await</code> adds quite a few things on top of this idea and adds some confusion with <code>await</code> keyword which doesn’t make current thread wait but works more like <a href="https://dkandalov.github.io/yielding-generators#basic-generator"><code>yield</code> in generators</a>.</p>
<p>Read next: <a href="https://dkandalov.github.io/call-with-current-continuation">call with current continuation</a>.</p>
</article>
</div>
</div> 
</div></div><br>
<br>
<br>
<br>
<br>
<br>
via dkandalov.github.io http://bit.ly/2F4Pv9U<br>
<br>
June 11, 2019 at 02:57PM
